### DISCUSSION TITLE
AuthProxy: Add support for auto_sign_up=false and invites when using auth proxy

### DISCUSSION THREAD

**Discussion by jameskirsop** [grafana/grafana#38368] [Dashboards, Panels & Visualizations]
**What happened**:
I invited a user, and they signed in via AzureAD hooked in to mod_auth_mellon which passed the user along to Grafana's auth_proxy and the following was logged in `Grafana.log`:
```
t=2020-03-05T15:17:53+1100 lvl=warn msg="Not allowing authproxy login, user not found in internal user database and allow signup = false"
t=2020-03-05T15:17:53+1100 lvl=eror msg="Failed to login" logger=auth.proxy username= message="Failed to log in as user, specified in auth proxy header" error="Invalid Username or Password"
t=2020-03-05T15:17:53+1100 lvl=eror msg="Failed to log in as user, specified in auth proxy header" logger=context error="Invalid Username or Password"
t=2020-03-05T15:17:53+1100 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/invite/bA3LMIZgwei4BAqmP490Oj7TXibGma status=407 remote_addr=114.141.104.83 time_ms=0 size=1761 referer=
```

**What you expected to happen**:
Given that the user was invited, they should exist in the Grafana database. Upon receiving the logon request, Grafana should move the user from 'Invited' to being an actual user.

**How to reproduce it (as minimally and precisely as possible)**:
Config:
```
[auth.anonymous]
enabled = false

[auth.proxy]
auto_sign_up = false
enabled = true

[users]
auto_assign_org = false
```

Try and log in with a username that matches one previously invited, via `auth_proxy`.

**Anything else we need to know?**:
The AzureAD integration works if I allow `auto_sign_up`, but the user gets _stuck_ in a new Org instead of placed in the one where the matching username is.
Invited users are placed in a particular Org for a purpose - and having users created and correctly configured *before* letting them know that they can sign in, is key for our use case.

As an aside, I'm also able to sign in with AzureAD for existing (full / not-pending invite) users without an issue.

**Environment**:
- Grafana version: 6.5.3
- Data source type & version: N/A
- OS Grafana is installed on: CentOS 8
- User OS & Browser: macOS, Chrome and Safari
- Grafana plugins: N/A
- Others:

**Comment by jameskirsop** [comment]
Updated to 6.6.2, issue still persists, but error messages have changed:
```
t=2020-03-06T09:15:52+1100 lvl=eror msg="Failed to get signed user" logger=auth.proxy username= message="User not found" error=nil
t=2020-03-06T09:15:52+1100 lvl=info msg="Request Completed" logger=context userId=0 orgId=0 uname= method=GET path=/ status=407 remote_addr=103.23.188.170 time_ms=1 size=1719 referer=https://login.microsoftonline.com/
```

**Comment by torkelo** [comment]
Invite feature is not designed to be used/paired with auth proxy, they need to complete the invite process (use the invite link) to create the user. 

For this scenario, you need to create the user first in the server admin section.

**Comment by jameskirsop** [comment]
@torkelo, I've tried that scenario. The downfall with that is the user gets created in their own Org, and trying to shuffle them into the correct one causes issues.

If I assign them to the correct org, then delete the 'users' org (by doing so I'm trying to get them automatically in the right one for the first time they sign in) - they can't log in because of a problem that isn't easily apparent from the debug logs.

In an ideal scenario, the user should never see their own organisation, and instead, always be dumped into the Organisation my admin user is currently browsing. This doesn't seem to be the current process flow of a new user being created in my testing.

Alternatively, provide an option on the new user form to select the organisation you wish to have them default too.

I've got a dev instance of Grafana running to try and work out what's going on here - but there's certainly a bug somewhere in the auth_proxy / user flow...

**Comment by jameskirsop** [comment]
> @torkelo, I've tried that scenario. The downfall with that is the user gets created in their own Org, and trying to shuffle them into the correct one causes issues.

Just to elaborate on this, when I create a new user via the admin interface with the above configuration the following procedure is done with negative results:

1. The user is placed into their own Organisation, named after their username
2. I place the user in the Organisation for their company
3. I remove the organisation that's been created with their name
4. This removes the association of the user to their 'own' org
5. At this point, the user is now in a single organisation
6. The user signs in
7. The user session is directed to a URL of `/?orgId=-1`, which doesn't exist.
8. The user is unable to view any details without editing the URL string (highly undesirable and our users consider this a bug)

What _should_ happen is that if a user is only in a single org, Grafana should direct the user to that single org, instead of the non-existant `-1`.

I'm not sure what should happen if a user is placed in multiple orgs - perhaps just direct them to the one with the lowest org ID and then from that point Grafana remembers what they viewed last?

It seems that I can work around this bug by sending the user to https://ourgrafanaurl.com/?orgId=4 (where 4 equals the correct id for the customer's organisation) for the first visit, but this is also less than ideal.

**Comment by pkolyvas** [comment]
We're in the process of moving Grafana feature requests to [GitHub Discussions](https://community.grafana.com/t/were-moving-grafana-feature-requests-to-github-discussions/50341/3). This issue is being moved to [Discussions](https://github.com/grafana/grafana/discussions) or has been moved from our issues list, depending on where you're seeing this message. 

 We hope to use the discussion thread to continue building use cases, and additional evidence based on engagement/voting, in order to help us prioritize this feature request later on down the road. 

As with all things Grafana, we value your input and participation in this project and appreciate your patience during this transitional period. Looking forward to the discussion!

**Comment by gelicia** [comment]
Hello, as you may have heard, we are [transitioning away from using discussions to discuss feature requests](https://github.com/grafana/grafana/discussions/73424). Due to the age and number of responses to this discussion, we are deciding to close it. If this is something you would like to see in Grafana, feel free to open an [issue](https://github.com/grafana/grafana/issues) so the discussion can continue. Thank you!
